### PR TITLE
enhance: Dendron still loads if there's a missing local vault

### PR DESCRIPTION
**Pull request by SeriousBug**
If there's a missing vault, instead of throwing an error and aborting the initialization, Dendron will now skip that vault and continue loading. This includes both the plugin and the CLI.

This is especially useful when publishing since the CLI does not clone missing remote vaults, which breaks publishing even if those vaults are not actually published.

Remove vaults that couldn't be cloned will still abort the initialization in the plugin. As part of self contained vaults, I'll later allow initialization to continue even if cloning a remote vault fails.

# Dendron Extended PR Checklist

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [x] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
  - Does not affect circular dependency count
- [x] sticking to existing conventions instead of creating new ones 
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended

- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [~] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [~] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed 
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics

- [~] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended

- [~] can we track the performance of this change to know if it is _successful_? 
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

## 

### Basics

- [~] [Write Tests](dev.process.qa.test) 
- [x] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [~] Common cases tested
- [~] 1-2 Edge cases tested
- [~] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended

- [~] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)

- [~] CSS
  - [ ] display is correct for following dimensions
    - [ ] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [ ] lg: screen ≥ 992px
    - [ ] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [ ] display is correct for following browsers (across the various dimensions)
    - [ ] safari
    - [ ] firefox
    - [ ] chrome

### Basics

- [~] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

## 

### Basics

- [~] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

## 

### Extended

- [~]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg. [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

File: packages/common-server/src/files.ts [hunk 1]
```diff
@@ -86,33 +91,51 @@ export function globMatch(patterns: string[] | string, fname: string): boolean {
...
+        })
+        .filter(isNotNull),
+      error: null,
+    };
+  } catch (err) {
+    return {
+      error: new DendronError({
+        message: "Error when reading the vault",
+        payload: err,
+        // Marked as minor to avoid stopping initialization. Even if we can't read one vault, we might be able to read other vaults.
```
Comment (jonathanyeung) [inline packages/common-server/src/files.ts:126]: Can you move this comment to the docstring of the method? That way it's easier for someone using it to know that this is the behavior.

File: packages/engine-server/src/drivers/file/storev2.ts [hunk 2]
```diff
@@ -401,13 +416,22 @@ export class FileStorage implements DStore {
         };
         const vpath = vault2Path({ vault, wsRoot: this.wsRoot });
         // OPT:make async and don't wait for return
-        if (!this.engine.config.noCaching) {
+        // Skip this if we found no notes, which means vault did not initialize
+        if (!this.engine.config.noCaching && notes.length > 0) {
```
Comment (jonathanyeung) [inline packages/engine-server/src/drivers/file/storev2.ts:420]: the comment doesn't seem to reflect the if statement? `notes.length > 0`

File: packages/plugin-core/src/commands/ReloadIndex.ts [hunk 3]
```diff
@@ -60,10 +61,20 @@ export class ReloadIndexCommand extends BasicCommand<
...
-    this.L.info({ ctx, vaultDir, msg: "creating root schema" });
-    await schemaModuleOpts2File(schema, vaultDir, "root");
-    return AutoFixAction.CREATE_ROOT_SCHEMA;
+    try {
+      const schema = SchemaUtils.createRootModule({ vault });
+      this.L.info({ ctx, vaultDir, msg: "creating root schema" });
+      await schemaModuleOpts2File(schema, vaultDir, "root");
+      return AutoFixAction.CREATE_ROOT_SCHEMA;
+    } catch (err) {
+      this.L.info({
```
Comment (jonathanyeung) [inline packages/plugin-core/src/commands/ReloadIndex.ts:70]: should this be error level?

**Review by jonathanyeung — CHANGES_REQUESTED** [review]

File: packages/engine-server/src/drivers/file/storev2.ts (hunk 2 above)
Comment (SeriousBug) [reply packages/engine-server/src/drivers/file/storev2.ts:420]: We only write notes to cache if there are notes to write, `notes.length > 0`. If `notes.length === 0`, then this block is skipped. Am I misreading the code?

File: packages/plugin-core/src/commands/ReloadIndex.ts (hunk 3 above)
Comment (SeriousBug) [reply packages/plugin-core/src/commands/ReloadIndex.ts:70]: This is expected if there are missing vaults. Should I log that at an error level?

File: packages/engine-server/src/drivers/file/storev2.ts (hunk 2 above)
Comment (jonathanyeung) [reply packages/engine-server/src/drivers/file/storev2.ts:420]: No, I'm being dumb 😀, please ignore.

File: packages/plugin-core/src/commands/ReloadIndex.ts (hunk 3 above)
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/ReloadIndex.ts:70]: Maybe keep at info for now.  Besides missing vaults, are there any other potential unexpected errors that may come out of the try block?

**Review by jonathanyeung — APPROVED** [review]

**Review by kevinslin — APPROVED** [review]

**Pull request merged**